File: src/main/java/com/app/umkaSchool/service/ProgressSnapshotService.java
package com.app.umkaSchool.service;

import com.app.umkaSchool.dto.progresssnapshot.ProgressSnapshotResponse;
import com.app.umkaSchool.model.ProgressSnapshot;
import com.app.umkaSchool.model.Student;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

public interface ProgressSnapshotService {
    /**
     * Create or update snapshot for today by incrementally adding session data.
     * Used when exercise attempt is completed.
     * Adds session data (from the completed attempt) to existing snapshot data.
     */
    ProgressSnapshot updateSnapshotAfterSession(Student student, com.app.umkaSchool.model.ExerciseAttempt completedAttempt);

    /**
     * Create or update snapshot for today. Used when exercise is completed.
     * @deprecated Use updateSnapshotAfterSession instead for incremental updates
     */
    @Deprecated
    ProgressSnapshot createOrUpdateTodaySnapshot(Student student);

    /**
     * Create snapshot for given date (used by scheduled job). If exists, updates it.
     */
    ProgressSnapshot createOrUpdateSnapshotForDate(Student student, LocalDate date);

    /**
     * Create snapshots for all students for today (scheduled job)
     */
    void createSnapshotsForAllStudentsToday();

    /**
     * Cleanup duplicate snapshots from yesterday: delete all except the last one if there were new attempts.
     * Runs at 1 AM to consolidate yesterday's snapshots.
     */
    void cleanupYesterdaySnapshots();

    /**
     * Get latest snapshot for a student
     */
    Optional<ProgressSnapshot> getLatestSnapshot(Student student);

    /**
     * Get snapshot for a specific date
     */
    Optional<ProgressSnapshot> getSnapshotForDate(Student student, LocalDate date);

    /**
     * Get all snapshots for a student
     */
    List<ProgressSnapshotResponse> getSnapshotsByStudentId(UUID studentId);

    /**
     * Get snapshots for a student within a date range
     */
    List<ProgressSnapshotResponse> getSnapshotsByDateRange(UUID studentId, LocalDate startDate, LocalDate endDate);

    /**
     * Get student statistics for a specific time period
     * @param student Student entity
     * @param period Time period: "day", "week", "month", or "all"
     * @return Statistics response with calculated metrics
     */
    com.app.umkaSchool.dto.stats.StatsResponse getStudentStats(Student student, String period);
}

File: frontend/src/services/progressSnapshot.service.ts
import axiosInstance from './axios.config';
import { ProgressSnapshot } from '../types/progress';

export const progressSnapshotService = {
    getLatestSnapshot: async (studentId: string): Promise<ProgressSnapshot | null> => {
        try {
            const response = await axiosInstance.get<ProgressSnapshot>(`/api/progress-snapshots/student/${studentId}/latest`);
            return response.data;
        } catch (error: any) {
            if (error.response?.status === 404) {
                return null; // No snapshot yet
            }
            console.error('Error fetching latest snapshot:', error);
            throw error;
        }
    },

    getSnapshotsByStudent: async (studentId: string): Promise<ProgressSnapshot[]> => {
        try {
            const response = await axiosInstance.get<ProgressSnapshot[]>(`/api/progress-snapshots/student/${studentId}`);
            return response.data;
        } catch (error) {
            console.error('Error fetching snapshots:', error);
            throw error;
        }
    },

    getSnapshotsByDateRange: async (
        studentId: string,
        startDate: string,
        endDate: string
    ): Promise<ProgressSnapshot[]> => {
        try {
            const response = await axiosInstance.get<ProgressSnapshot[]>(
                `/api/progress-snapshots/student/${studentId}/date-range`,
                {
                    params: { startDate, endDate }
                }
            );
            return response.data;
        } catch (error) {
            console.error('Error fetching snapshots by date range:', error);
            throw error;
        }
    },

    getSnapshotByDate: async (studentId: string, date: string): Promise<ProgressSnapshot | null> => {
        try {
            const response = await axiosInstance.get<ProgressSnapshot>(
                `/api/progress-snapshots/student/${studentId}/date/${date}`
            );
            return response.data;
        } catch (error: any) {
            if (error.response?.status === 404) {
                return null;
            }
            console.error('Error fetching snapshot by date:', error);
            throw error;
        }
    }
};

